File: restapi/client/src/main/kotlin/io/connorwyatt/todos/restapi/client/TodosClient.kt
package io.connorwyatt.todos.restapi.client

import io.connorwyatt.common.http.EmptyHttpResult
import io.connorwyatt.common.http.HttpResult
import io.connorwyatt.todos.restapi.models.Todo
import io.connorwyatt.todos.restapi.models.TodoDefinition
import io.connorwyatt.todos.restapi.models.TodoPatch
import io.connorwyatt.todos.restapi.models.TodoReference

interface TodosClient {
    suspend fun retrieveTodos(): HttpResult<List<Todo>>

    suspend fun retrieveTodo(todoId: String): HttpResult<Todo>

    suspend fun addTodo(definition: TodoDefinition): HttpResult<TodoReference>

    suspend fun updateTodo(todoId: String, patch: TodoPatch): EmptyHttpResult

    suspend fun completeTodo(todoId: String): EmptyHttpResult
}

File: ui/src/todos/api/todos-client.ts
import { parseJSON } from "date-fns"

import { httpClient } from "~/shared/api/http/http-client"
import { Todo } from "~/todos/api/models/todo"
import { TodoDefinition } from "~/todos/api/models/todo-definition"
import { TodoReference } from "~/todos/api/models/todo-reference"
import { TodoResponse } from "~/todos/api/models/todo-response"

const getTodos = async (): Promise<Todo[]> => {
    const response = await httpClient.get<TodoResponse[]>("/api/todos")

    return response.data.map(mapTodo)
}

const getTodo = async (todoId: string): Promise<Todo> => {
    const response = await httpClient.get<TodoResponse>(`/api/todos/${todoId}`)

    return mapTodo(response.data)
}

const addTodo = async (todoDefinition: TodoDefinition): Promise<TodoReference> => {
    const response = await httpClient.post<TodoReference>(`/api/todos`, todoDefinition)

    return response.data
}

const completeTodo = async (todoId: string): Promise<void> => {
    await httpClient.post(`/api/todos/${todoId}/actions/complete`)
}

export const todosClient = {
    getTodos,
    getTodo,
    addTodo,
    completeTodo,
}

const mapTodo = ({ addedAt, completedAt, ...rest }: TodoResponse): Todo => ({
    ...rest,
    addedAt: parseJSON(addedAt),
    completedAt: completedAt != null ? parseJSON(completedAt) : null,
})
